File: modules/widgetkit/ios/WidgetKitModule.swift
import ExpoModulesCore
import WidgetKit

public class WidgetKitModule: Module {
    let _groupName: String = "group.com.pourtainer.mobile"
    let _connectionsKey: String = "pourtainer::connections"
	let _isSubscribedKey: String = "pourtainer::subscribed"

    private func getConnections() -> [Connection] {
        guard let sharedDefaults = UserDefaults(suiteName: _groupName),
              let rawExistingConnections = sharedDefaults.data(forKey: _connectionsKey) else {
            return []
        }

        return (try? JSONDecoder().decode([Connection].self, from: rawExistingConnections)) ?? []
    }
	
	private func reloadWidgets() {
        if #available(iOS 16.0, *) {
            WidgetCenter.shared.invalidateConfigurationRecommendations()
        }
        
        WidgetCenter.shared.reloadAllTimelines()
    }

    public func definition() -> ModuleDefinition {
        Name("PourtainerWidgetKit")
		
		Function("setIsSubscribed") { (isSubscribed: Bool) -> Void in
            guard let sharedDefaults = UserDefaults(suiteName: _groupName) else { return }
            
            sharedDefaults.set(isSubscribed, forKey: _isSubscribedKey)
            
            self.reloadWidgets()
        }

        Function("getConnections") {
            return self.getConnections()
        }

        Function("addConnection") { (connection: Connection) in
            guard let sharedDefaults = UserDefaults(suiteName: _groupName) else {
                return
            }

            do {
                var connections = self.getConnections()

                // add or update Connection
                if let index = connections.firstIndex(where: { $0.id == connection.id }) {
                    connections[index] = connection
                } else {
                    connections.append(connection)
                }

                let encodedConnections = try JSONEncoder().encode(connections)

                sharedDefaults.set(encodedConnections, forKey: _connectionsKey)
                sharedDefaults.synchronize()

                 self.reloadWidgets()
            } catch {
                // for now do nothing
            }
        }
		
		Function("removeConnection") { (id: String) in
            guard let sharedDefaults = UserDefaults(suiteName: _groupName) else { return }
            
            do {
                let connections = self.getConnections().filter { $0.id != id }
                let encodedConnections = try JSONEncoder().encode(connections)
                
                sharedDefaults.set(encodedConnections, forKey: _connectionsKey)
                
                self.reloadWidgets()
            }
        }

        Function("clearAllConnections") {
            guard let sharedDefaults = UserDefaults(suiteName: _groupName) else {
                return
            }

            sharedDefaults.removePersistentDomain(forName: _groupName)
            sharedDefaults.synchronize()

            self.reloadWidgets()
        }
    }
}

File: modules/widgetkit/android/src/main/java/expo/modules/widgetkit/WidgetKitModule.kt
package expo.modules.widgetkit

import android.content.Context
import android.content.Intent
import com.google.gson.Gson
import expo.modules.kotlin.modules.Module
import expo.modules.kotlin.modules.ModuleDefinition
import androidx.core.content.edit

class WidgetKitModule : Module() {
    companion object {
        const val groupName = "group.com.pourtainer.mobile"
        const val connectionsKey = "pourtainer::connections"
		const val isSubscribedKey = "pourtainer::subscribed"
    }

    private fun notifyAllWidgets() {
        val intent = Intent().apply {
            action = "android.appwidget.action.APPWIDGET_UPDATE"
        }

        appContext.reactContext?.sendBroadcast(intent)
    }

    private fun getConnections(): List<Connection> {
        val rawConnections = appContext.reactContext
            ?.getSharedPreferences(groupName, Context.MODE_PRIVATE)
            ?.getString(connectionsKey, "[]")

        return rawConnections?.let {
            Gson().fromJson(it, Array<Connection>::class.java).toList()
        } ?: emptyList()
    }

    override fun definition() = ModuleDefinition {
        Name("PourtainerWidgetKit")

        Function("getConnections") {
            return@Function this@WidgetKitModule.getConnections().map {
                mapOf(
                    "accessToken" to it.accessToken,
                    "id" to it.id,
                    "url" to it.url,
                )
            }
        }
		
		Function("setIsSubscribed") { isSubscribed: Boolean ->
            appContext.reactContext?.getSharedPreferences(groupName, Context.MODE_PRIVATE)?.let { prefs ->
                prefs.edit() {
                    putBoolean(isSubscribedKey, isSubscribed)
                    apply()
                }

                notifyAllWidgets()
            }
        }

        Function("addConnection") { connection: Connection ->
            appContext.reactContext?.getSharedPreferences(groupName, Context.MODE_PRIVATE)?.let { prefs ->
                val connections = this@WidgetKitModule.getConnections().toMutableList()

                // Add or update connection
                val index = connections.indexOfFirst { it.id == connection.id }

                if (index != -1) {
                    connections[index] = connection
                } else {
                    connections.add(connection)
                }

                prefs.edit() {
                    putString(connectionsKey, Gson().toJson(connections))
                    apply()
                }

                notifyAllWidgets()
            }
        }
		
		Function("setConnections") { connections: List<Connection> ->
			appContext.reactContext?.getSharedPreferences(groupName, Context.MODE_PRIVATE)?.let { prefs ->
                prefs.edit() {
                    putString(connectionsKey, Gson().toJson(connections))
                    apply()
                }

                notifyAllWidgets()
			}
        }
		
		Function("removeConnection") { id: String ->
            appContext.reactContext?.getSharedPreferences(groupName, Context.MODE_PRIVATE)?.let { prefs ->
                val connections = this@WidgetKitModule.getConnections().toMutableList().filter { it.id != id}

                prefs.edit() {
                    putString(connectionsKey, Gson().toJson(connections))
                    apply()
                }

                notifyAllWidgets()
            }
        }

		// also clears isSubscribed
        Function("clearAllConnections") {
            appContext.reactContext?.getSharedPreferences(groupName, Context.MODE_PRIVATE)?.let { prefs ->
                prefs.edit() {
                    clear()
                    apply()
                }

                notifyAllWidgets()
            }
        }
    }
}
